### ISSUE TITLE
NBT exception while reading levels db

### ISSUE THREAD

**Issue by cabbey** [bedrock-viz/bedrock-viz#47]
Working from master (I pulled fresh this evening and got a97c0e2b657594c95a1ccf011fa9cc8cc52de8d1) I'm getting this error collection while parsing my save files.
```
[  error] NBT exception: (Invalid tag type: 0) (eof=false) (is=false) (tc=144) (pos=-1) (buflen=13563) (parseNbtQuiet)
[warning] Unexpected NBT format in _do_chunk_v7
[  error] NBT exception: (Invalid tag type: 0) (eof=false) (is=false) (tc=137) (pos=-1) (buflen=13237) (parseNbtQuiet)
[warning] Unexpected NBT format in _do_chunk_v7
```
It occurs twice in the output, once during the parse all level db step and again during the output for the overworld, and there are chunks that are missing in the resulting render which I know I've explored many times (this one cuts across the trade road from my home base to a village):

![image](https://user-images.githubusercontent.com/522147/86891917-64711400-c0b4-11ea-8b50-0ec103d99b89.png)

and which I *think* I've explored, but can't be sure:

![image](https://user-images.githubusercontent.com/522147/86892190-c5005100-c0b4-11ea-99a4-fe969630c713.png)

This is on an Ubuntu 19.10 box running the dedicated server. The [scripts I use to manage the dedicated server](https://github.com/TapeWerm/MCscripts) does a nightly backup at 4am and I just pull those off my NAS and unzip them locally to parse, then write the output back to the SMB mount which is then served up via nginx. (Once I figured out that the leveldb library can't cope with opening the database over an SMB mount at least... that took a couple hours.)

Full output dump:
[run errors.txt](https://github.com/bedrock-viz/bedrock-viz/files/4889207/run.errors.txt)

There are also a bunch of weird orphaned chunks spread all around the map edges which make no sense to me, I'm not sure if that's related. 
![image](https://user-images.githubusercontent.com/522147/86893347-58865180-c0b6-11ea-9aa5-bb5051c129b5.png)

**Comment by jasper-wan** [comment]
do you mind sharing your world save with us? It would be helpful if we can test it locally

**Comment by jasper-wan** [comment]
and which version of dedicated server are you using?

**Comment by cabbey** [comment]
The scripts I use automatically keep the server running whatever is "current" from MS... looks like it's 1.16.1.2 currently (and was at the time the backup was made).

I'm hesitant to share the world save publicly, as my son and I use this world for achievements and I don't know how to mask the world id that MS uses to invalidate worlds that have had cheats turned on. I've already accidentally ruined one world we used to use for that by bringing up a copy on a separate machine with cheats enabled to get coordinates turned on so I could calculate the angle to build a tunnel to make a straight line between two places... I'd really rather not risk that happening again. That said, I'd be willing to share it with individuals working on this codebase directly by email if they'll agree to keep it private. I can be reached at my nickname at my nickname dot net. :)

**Comment by jasper-wan** [comment]
I've send you an email. The save will be used for development only and won't be loaded by the official server by MS.

**Comment by cabbey** [comment]
I've sent it along, and can confirm the next day's backup has the same issue.

**Comment by cabbey** [comment]
I should note, I'm on an x86_64 system, and there are a couple compile warnings, including in the nbt code, during the build... not sure if they're related as I've not gone too far into the codebase.

```
/home/cabbey/bedrock-viz/src/main.cc: In function ‘int32_t mcpe_viz::doParseConfigFile(const string&)’:
/home/cabbey/bedrock-viz/src/main.cc:249:18: warning: ignoring return value of ‘char* fgets(char*, int, FILE*)’, declared with attribute warn_unused_result [-Wunused-result]
             fgets(buf, 1024, fp);
             ~~~~~^~~~~~~~~~~~~~~
/home/cabbey/bedrock-viz/src/nbt.cc: In member function ‘std::__cxx11::string mcpe_viz::ParsedVillageVillager::toGeoJSON()’:
/home/cabbey/bedrock-viz/src/nbt.cc:2487:32: warning: format ‘%lld’ expects argument of type ‘long long int’, but argument 3 has type ‘int64_t’ {aka ‘long int’} [-Wformat=]
             sprintf(tmpstring, "\"id\":%lld", id);
                                ^~~~~~~~~~~~~  ~~
/home/cabbey/bedrock-viz/src/world/world.cc: In member function ‘int32_t mcpe_viz::MinecraftWorld_LevelDB::parseLevelFile(const string&)’:
/home/cabbey/bedrock-viz/src/world/world.cc:86:14: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’, declared with attribute warn_unused_result [-Wunused-result]
         fread(&fVersion, sizeof(int32_t), 1, fp);
         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/cabbey/bedrock-viz/src/world/world.cc:87:14: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’, declared with attribute warn_unused_result [-Wunused-result]
         fread(&bufLen, sizeof(int32_t), 1, fp);
         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/cabbey/bedrock-viz/src/world/world.cc:95:18: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’, declared with attribute warn_unused_result [-Wunused-result]
             fread(buf, 1, bufLen, fp);
             ~~~~~^~~~~~~~~~~~~~~~~~~~
In file included from /home/cabbey/bedrock-viz/src/world/world.cc:1:
/home/cabbey/bedrock-viz/src/world/world.h: In member function ‘int32_t mcpe_viz::MinecraftWorld_LevelDB::parseLevelName(const string&)’:
/home/cabbey/bedrock-viz/src/world/world.h:82:18: warning: ignoring return value of ‘char* fgets(char*, int, FILE*)’, declared with attribute warn_unused_result [-Wunused-result]
             fgets(buf, 1024, fp);
             ~~~~~^~~~~~~~~~~~~~~
```

The above from a clean build of master @ eaf4676719b1762b1a9fd9b8fbabfa45a7298c25 after installing the new boost pre-req.

**Comment by tomnolan** [comment]
Just as a note regarding the "orphaned" chunks in the map. I don't believe this is actually an error based on what I've seen before, I've also had this happen. I'm pretty sure this has something to do with "special" chunks that bedrock will load when you are in-game that are beyond the normal chunk render distance, but due to the nature of what those chunks represent, it loads them anyway. If you were to go exploring a bit further out and compare renders of the maps, you should see that those chunks are actually correctly aligned with the world when you load the surrounding chunks. I'm pretty sure there is a comment about this somewhere in the wiki, I'll have to see if I can find it.

**Comment by cabbey** [comment]
@tomnolan indeed. Turning on the overlays for items shows every one of them has some collection of chests and spawners in them... and exploring out to a couple of them didn't add anything of interest in the neighboring chunks.

**Comment by cabbey** [comment]
I have removed two chunks of what now appear to be unrelated information from the main report. The orphaned blocks on the edges contain things the server has pre-created out and the two "broken" strips have gone away in later runs, but the errors as above remain. (I refresh the map based on a nightly backup before we start playing if we're going to use it as a reference in that session.)
